### ISSUE TITLE
Re-think the `SSAValue.name` field

### ISSUE THREAD

**Issue by AntonLydike** [xdslproject/xdsl#571]
We currently impose the restriction on the `name` field that it cannot end in a number, because of complicated reasons that boil down to the assumption that no `SSAValue`s may ever have the same `.name` field.

I've thought about it for a bit, and I think that requirement is nonsense. We can have SSA Values named whatever the heck we want, even all the same if we want to. The printer should make sure that no SSA value names overlap in the printed program. If we impose that restriction earlier it makes no sense at all, as it really isn't enforceable.

If we replace an operation with multiple ops that compute the value, then we can just keep the name for all intermediate values, and the printer will add `_0..n` suffixes. That's much easier!

Am I overlooking something, or can we implement this?

**Comment by webmiche** [comment]
I think we can do this. In particular, actual names for SSAValues mostly make sense in IR that is actually introspectable, so it won't be very big either way, so low possibility for clashes.

What I don't like very much is that this makes our IRs potentially not roundtrip anymore, so the printing and parsing the data structure again might not be the same data structure anymore. Though, I am not even 100% confident that this is the case atm, so that's a small contrapoint...

**Comment by superlopuh** [comment]
It seems like compatibility with MLIR should trump aesthetics for the foreseeable future, even though it definitely seems like an unfortunate restriction to me.

**Comment by math-fehr** [comment]
Yeah sure, we can remove that restriction.
The only reason why there was this restriction is that we wanted to write quickly something simple, we didn't want to overthink this, so we took the easy way out.

Otherwise, our IRs already don't round-trip (at least in the sense that wouldn't print themselves exactly the same), so it won't change much imo.

I'm not sure we care that much about being 100% compatible with MLIR naming-wise, in the sense that it doesn't really matter if our output is 100% the same, since those names are just indications anyway. As long as we keep the same name base or something, that's fine.

**Comment by superlopuh** [comment]
It would be quite powerful if we could guarantee exact compatibility, there are some cases where diffing the xdsl and mlir outputs has been useful, and to debug I had to manually adjust the differences. Especially as we port more of the functionality of MLIR, if we could iron out the differences little by little that would be great, I have a feeling it wouldn't take that much effort to get there if we do this incrementally.

**Comment by superlopuh** [comment]
But back to this issue, if MLIR doesn't have that restriction then it seems like a great one to remove in xDSL also

**Comment by AntonLydike** [comment]
I still don't see how changing the SSA value naming would make this worse. Relaxing restrictions here (without producing invalid output) can only improve our ability to replicate MLIR output. Or am I missing something here?

**Comment by webmiche** [comment]
No, I think youa re right. The point I was going for was that two equally names SSAValues won't be equally names after parsing again. But if you have SSAValues with equal names, you are already lost either way (at least in the same scope).

IMO, getting this consistent down to the level of printing the same number feels just painful to me. I would prefer implementing our own version of diff, that tracks that the def-use chains align and does not check for exact equality.

**Comment by AntonLydike** [comment]
I managed to implement this without much effort, it also seems to not affect the filechecks.

I changed the renaming scheme for overlapping names to `_{n}` instead of just appending n as I think that's how MLIR does it as well (I think).
